src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -1278,6 +1288,30 @@ void ZooKeeper::receiveEvent()
             response->removeRootPath(root_path);
         }
 
+        /// Instead of setting the watch in sendEvent, set it in receiveEvent becuase need to check the response.
```

Comment:
Ok. It looks correct *under assumption that the responses from ZooKeeper are received in order*.

For example, if we have received a response for a command where a watch was set, then we cannot receive the corresponding watch event earlier.

Are you sure that this assumption is correct?